### PR TITLE
Integrals

### PR DESCRIPTION
**Pull request by rbeucher**
I would do this instead @lmoresi @julesghub 
That's the pattern I see in the Petsc examples...

### CODE REVIEW

**Comment by rbeucher** [comment]
What do you think?

**Comment by lmoresi** [comment]
I'm not sure (well, except, can we not have those vtk files !) 

1) Is this the same pattern you would use for the solvers ?

2) I'd like to have to old way deprecated but not removed while we test things as we are now starting to use the code for research problems.

**Comment by rbeucher** [comment]
Sorry, the vtk files were not supposed to be there...

**Comment by rbeucher** [comment]
Well, I would like to discuss this with @julesghub and @jmansour . Not sure about the solver but that story with the quadrature seems odd to me.

File: underworld3/maths.pyx [hunk 1]
```diff
@@ -71,36 +73,24 @@ class Integral:
 
         # Pull out vec for variables, and go ahead with the integral
         self.mesh.update_lvec()
-        a_global = self.mesh.dm.getGlobalVec()
-        self.mesh.dm.localToGlobal(self.mesh.lvec, a_global)
-        cdef Vec cgvec
-        cgvec = a_global
+        cdef Vec cgvec = self.mesh.dm.createGlobalVec()
```
Comment (jmansour) [inline underworld3/maths.pyx]: This will be a hefty malloc for every call.  Why change from the `getGlobalVec()` routine?

Comment (knepley) [reply underworld3/maths.pyx]: Same question

File: underworld3/maths.pyx [hunk 2]
```diff
@@ -71,36 +73,24 @@ class Integral:
...
         # Now set callback... note that we set the highest degree var_id (as determined
         # above) for the second parameter. 
-        ierr = PetscDSSetObjective(ds.ds, 0, ext.fns_residual[0]); CHKERRQ(ierr)
+        CHKERRQ( PetscDSSetObjective(ds.ds, 0, ext.fns_residual[0]) )
         
         cdef PetscScalar val
-        ierr = DMPlexComputeIntegralFEM(dm.dm, cgvec.vec, <PetscScalar*>&val, NULL); CHKERRQ(ierr)
-        self.mesh.dm.restoreGlobalVec(a_global)
+        CHKERRQ( DMPlexComputeIntegralFEM(dmc.dm, cgvec.vec, <PetscScalar*>&val, NULL) )
+        cgvec.destroy()
```
Comment (knepley) [inline underworld3/maths.pyx]: I would put back the Restore here

Comment (rbeucher) [reply underworld3/maths.pyx]: Yes. I will. Thanks.

File: underworld3/maths.pyx [hunk 3]
```diff
@@ -71,36 +73,21 @@ class Integral:
...
+        dmc.setField(0, fec)
+        dmc.createDS()
+
+        cdef DS ds = dmc.getDS()
+        CHKERRQ( PetscDSSetObjective(ds.ds, 0, ext.fns_residual[0]) )
         
         cdef PetscScalar val
-        ierr = DMPlexComputeIntegralFEM(dm.dm, cgvec.vec, <PetscScalar*>&val, NULL); CHKERRQ(ierr)
-        self.mesh.dm.restoreGlobalVec(a_global)
+        CHKERRQ( DMPlexComputeIntegralFEM(dmc.dm, cgvec.vec, <PetscScalar*>&val, NULL) )
```
Comment (jmansour) [inline underworld3/maths.pyx:89]: I'm a bit confused about how this comes together.

You've taken a clone (`dmc`) of the original DM, added a new FE, and set it as field `0` on `dmc`.

Does the data in `cgvec` (from the original DM) map correctly within the `dmc` layout? At the very least, isn't field `0` potentially different order on the DMs? 

Does this work for anything more complex than a single field DM?

Comment (rbeucher) [reply underworld3/maths.pyx:89]: Yeah, I see what you mean. I'm using the functions to calculate cell areas and that's fine for that but it's rather fragile for anything else...

Comment (knepley) [reply underworld3/maths.pyx:89]: Yes, I should have better documentation here. DM is doing too much (I think). It is the topology, geometry, function spaces, and equations.

When you DMClone() you keep the same topology and geometry, but discard the function spaces and equations. So here, you attach a new space and a function you want to integrate over the mesh (called, perhaps unfairly, the objective).

 I think this pattern is fine, and you should not have problems. I tend to clone a new DM for each new operation or integration I want to do.

**Comment by julesghub** [comment]
Closing as this is old and "working now"

**Pull request closed without merging**